### PR TITLE
Throttle RC actions

### PR DESCRIPTION
**Pull request by Andyz26**
### Context

Add rate-limit to resource cluster actor backed APIs so that the mailbox backlog is protected from repeated calls.
E.g. TE 1 request registration -> adds message to RC actor mailbox -> request timed out on API ask -> deadletter returned by RC actor  -> TE request again due to previous timeout. 

### Checklist

- [x] `./gradlew build` compiles code correctly
- [x] Added new tests where applicable
- [x] `./gradlew test` passes all tests
- [x] Extended README or added javadocs where applicable

### CODE REVIEW

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java [hunk 1]
```diff
@@ -19,37 +19,69 @@
...
+            ResourceClusterTaskExecutorMapper mapper,
+            int maxConcurrentRequestCount) {
+        this.resourceClusterManagerActor = resourceClusterManagerActor;
+        this.askTimeout = askTimeout;
+        this.mapper = mapper;
+
+        this.rateLimiter = RateLimiter.create(maxConcurrentRequestCount);
+    }
+
+    protected <In, Out> Out withThrottle(Function<In, Out> func, In req) throws RequestThrottledException {
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:53]: nit: make this private.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java [hunk 2]
```diff
@@ -19,37 +19,69 @@
...
+        if (this.rateLimiter.tryAcquire(2, TimeUnit.SECONDS)) {
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:54]: nit: 2 seconds looks too high.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java (hunk 1 above)
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:53]: I would probably decorate the function rather than apply it here. 
```
private <In, Out> Function<In, CompletableFuture<Out>> withThrottle(Function<In, CompletableFuture<Out>> func) {
  return in -> {
    if (rateLimiter.tryAcquire(timeout)) {
      return func.apply(in);
    } else {
      return CompletableFutures.exceptionallyCompletedFuture(new RequestThrottleException(in.getClass()));
    }
  };
}
```

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManager.java [hunk 3]
```diff
@@ -59,6 +59,9 @@ interface ExecutorStateManager {
     @Nullable
     TaskExecutorState get(TaskExecutorID taskExecutorID);
 
+    @Nullable
+    TaskExecutorState getIncludeArchived(TaskExecutorID taskExecutorID);
+
```
[the remark this answers is not in the document]
Comment (Andyz26) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ExecutorStateManager.java:64]: Fixing other issues seen in the snapshot build.

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java (hunk 1 above)
Comment (Andyz26) [reply mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:53]: might need to apply in the inheritor.

File: mantis-server/mantis-server-agent/src/test/java/io/mantisrx/server/agent/ResourceManagerGatewayCxnTest.java [hunk 4]
```diff
@@ -134,7 +135,7 @@ public CompletableFuture<Void> answer(InvocationOnMock invocation) {
     }
 
     @Test
-    public void testWhenHeartbeatFailsContinuously() {
+    public void testWhenHeartbeatFailsContinuously() throws RequestThrottledException {
```
Comment (sundargates) [inline mantis-server/mantis-server-agent/src/test/java/io/mantisrx/server/agent/ResourceManagerGatewayCxnTest.java:138]: I'm guessing this change is not needed?

File: mantis-server/mantis-server-agent/src/test/java/io/mantisrx/server/agent/RuntimeTaskImplExecutorTest.java [hunk 5]
```diff
@@ -117,7 +118,7 @@ public class RuntimeTaskImplExecutorTest {
     private CollectingTaskLifecycleListener listener;
 
     @Before
-    public void setUp() throws IOException {
+    public void setUp() throws IOException, RequestThrottledException {
```
Comment (sundargates) [inline mantis-server/mantis-server-agent/src/test/java/io/mantisrx/server/agent/RuntimeTaskImplExecutorTest.java:121]: same here.

File: mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/TaskExecutor.java [hunk 6]
```diff
@@ -412,8 +412,8 @@ public void runOneIteration() throws Exception {
                     .thenComposeAsync(report -> {
                         log.debug("Sending heartbeat to resource manager {} with report {}", gateway, report);
                         return gateway.heartBeatFromTaskExecutor(
-                            new TaskExecutorHeartbeat(taskExecutorRegistration.getTaskExecutorID(),
-                                taskExecutorRegistration.getClusterID(), report));
+                                new TaskExecutorHeartbeat(taskExecutorRegistration.getTaskExecutorID(),
+                                        taskExecutorRegistration.getClusterID(), report));
```
Comment (sundargates) [inline mantis-server/mantis-server-agent/src/main/java/io/mantisrx/server/agent/TaskExecutor.java:416]: nit: revert these changes?

File: mantis-control-plane/mantis-control-plane-server/src/test/java/io/mantisrx/master/resourcecluster/ResourceClusterActorTest.java [hunk 7]
```diff
@@ -427,7 +429,7 @@ public void testAssignmentTimeout() throws Exception {
     }
 
     @Test
-    public void testGetMultipleActiveJobs() throws ExecutionException, InterruptedException {
+    public void testGetMultipleActiveJobs() throws ExecutionException, InterruptedException, RequestThrottledException {
```
Comment (sundargates) [inline mantis-control-plane/mantis-control-plane-server/src/test/java/io/mantisrx/master/resourcecluster/ResourceClusterActorTest.java:432]: I'm guessing this is also not needed.

**Review by sundargates — APPROVED** [review]

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java [hunk 8]
```diff
@@ -18,38 +18,73 @@
...
+    ResourceClusterGatewayAkkaImpl(
+            ActorRef resourceClusterManagerActor,
+            Duration askTimeout,
+            ResourceClusterTaskExecutorMapper mapper,
+            int maxConcurrentRequestCount) {
+        this.resourceClusterManagerActor = resourceClusterManagerActor;
+        this.askTimeout = askTimeout;
+        this.mapper = mapper;
+
+        this.rateLimiter = RateLimiter.create(maxConcurrentRequestCount);
```
Comment (fdc-ntflx) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:51]: nit: s/maxConcurrentRequestCount/maxPermitsPerSecond

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java [hunk 9]
```diff
@@ -18,38 +18,73 @@
...
+    }
+
+    private <In, Out> Function<In, CompletableFuture<Out>> withThrottle(Function<In, CompletableFuture<Out>> func) {
```
Comment (fdc-ntflx) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:54]: not a blocker, but have you thought about making this an interface all our routes could implement/use?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java [hunk 10]
```diff
@@ -18,38 +18,73 @@
...
+        return in -> {
+            if (rateLimiter.tryAcquire(1, TimeUnit.SECONDS)) {
```
Comment (fdc-ntflx) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:56]: nit: make this configurable?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/config/MasterConfiguration.java [hunk 11]
```diff
@@ -378,6 +378,10 @@ default Duration getSchedulerIntervalBetweenRetries() {
     @Default("")
     String getJobClustersWithArtifactCachingEnabled();
 
+    // rate limit actions on resource cluster actor to control backlog.
+    @Config("mantis.master.resource.cluster.actions.permitsPerSecond")
+    @Default("300")
```
Comment (fdc-ntflx) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/server/master/config/MasterConfiguration.java:383]: curious, why 300?

File: mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java (hunk 8 above)
Comment (fdc-ntflx) [inline mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterGatewayAkkaImpl.java:51]: Is the rate limiter global for all actions? Have you considered separating it by action? I'm thinking about the case where a single action fills up the limit impeding all other actions to be performed.

**Review by fdc-ntflx — APPROVED** [review]
